src/virtual-machine.js

```diff
@@ -542,17 +544,59 @@ class VirtualMachine extends EventEmitter {
     /**
      * Get an SVG string from storage.
      * @param {int} costumeIndex - the index of the costume to be got.
-     * @return {string} the costume's SVG string, or null if it's not an SVG costume.
+     * @return {string|HTMLImageElement} the costume's SVG string if it's SVG,
+     *     an HTMLImageElement if it's a PNG, or null if it couldn't be found or decoded.
      */
-    getCostumeSvg (costumeIndex) {
+    getCostume (costumeIndex) {
         const id = this.editingTarget.getCostumes()[costumeIndex].assetId;
-        if (id && this.runtime && this.runtime.storage &&
-                this.runtime.storage.get(id).dataFormat === 'svg') {
+        if (!id || !this.runtime || !this.runtime.storage) return null;
+        const format = this.runtime.storage.get(id).dataFormat;
+        if (format === this.runtime.storage.DataFormat.SVG) {
             return this.runtime.storage.get(id).decodeText();
+        } else if (format === this.runtime.storage.DataFormat.PNG) {
+            const data = this.runtime.storage.get(id).encodeDataURI('image/png');
+            const image = new Image();
+            image.src = data;
+            return image;
         }
+        log.error(`Unhandled format: ${this.runtime.storage.get(id).dataFormat}`);
         return null;
     }
 
+    /**
+     * Update a costume with the given bitmap
+     * @param {int} costumeIndex - the index of the costume to be updated.
+     * @param {HTMLCanvasElement} bitmap - new bitmap for the renderer.
+     * @param {number} rotationCenterX x of point about which the costume rotates, relative to its upper left corner
+     * @param {number} rotationCenterY y of point about which the costume rotates, relative to its upper left corner
+     */
+    updateBitmap (costumeIndex, bitmap, rotationCenterX, rotationCenterY) {
+        const costume = this.editingTarget.getCostumes()[costumeIndex];
+        if (costume && this.runtime && this.runtime.renderer) {
+            costume.rotationCenterX = rotationCenterX;
+            costume.rotationCenterY = rotationCenterY;
+            this.runtime.renderer.updateBitmapSkin(costume.skinId, bitmap, [rotationCenterX, rotationCenterY]);
+        }
+
+        bitmap.toBlob(blob => {
+            const reader = new FileReader();
+            reader.addEventListener('loadend', () => {
+                const storage = this.runtime.storage;
+                costume.assetId = storage.builtinHelper.cache(
+                    storage.AssetType.ImageBitmap,
+                    storage.DataFormat.PNG,
+                    Buffer.from(reader.result)
+                );
+                // @todo is there a better way to make sure all info is up to date on the costume?
```

Comment:
I think you can use `bitmap.getContext('2d').getImageData(0, 0, width, height).data` where width and height are the width and height of the expected image -- instead of calling toBlob and using a FileReader. At any rate, instead of using `Buffer.from` above, you can just use `new Uint8Array` above.

Reply:
Thanks!! That's much easier

Reply:
Here are the data URIs that come out in case that helps in figuring out what went wrong. The one coming from imagedata.data seems to be infested with 'A's and forward slashes.

[imagedatadata.txt](https://github.com/LLK/scratch-vm/files/1961189/imagedatadata.txt)
[readerbufferdata.txt](https://github.com/LLK/scratch-vm/files/1961191/readerbufferdata.txt)

Reply:
@fsih I think `<all that canvas context stuff>.data.buffer` will work. 
If you're getting the uri anyways, you can use canvas.toDataURL('image/png') instead.

Reply:
imageData.data.buffer isn't working for me either u_u the entire returned dataURI is "data:image/png;base64"

Every method I've tried so far via image data has given me back an unloadable dataURI including:
Buffer.from(imagedata.data)
new Uint8Array(imagedata.data)
imagedata.data